planner/core/common_plans.go

```diff
@@ -375,6 +376,12 @@ REBUILD:
 	e.Plan = p
 	_, isTableDual := p.(*PhysicalTableDual)
 	if !isTableDual && prepared.UseCache && !stmtCtx.OptimDependOnMutableConst {
+		// rebuild key to exclude kv.TiFlash when stmt is not read only
+		if _, isolationReadContainTiFlash := sessVars.IsolationReadEngines[kv.TiFlash]; isolationReadContainTiFlash && !IsReadOnly(stmt, sessVars) {
+			delete(sessVars.IsolationReadEngines, kv.TiFlash)
+			cacheKey = NewPSTMTPlanCacheKey(sctx.GetSessionVars(), e.ExecID, prepared.SchemaVersion)
+			sessVars.IsolationReadEngines[kv.TiFlash] = struct{}{}
+		}
```

Comment:
When use binary protocol and the plan cache is not hit, a new cache will generate inside `getPhysicalPlan` function, in which use the following codes to exclude `TiFlash` engine for write statement.
https://github.com/pingcap/tidb/blob/2364fec8f04710db0b1ec24be6f9e9bdb93ced5b/planner/optimize.go#L85-L88
After the plan is builing complete, the `TiFlash` engine is set back and the it'll still write a plan-cache key with `TiFlash` engine. As a result, we cache a write plan for `TiFlash` engine, and it'll never be used correctly.

Reply:
:open_mouth: @winoros  PTAL